### PR TITLE
DOC: Migration guide: Deferred task creation

### PR DESCRIPTION
**Pull request by anton-potapov**
- added Deferred task creation chapter

Signed-off-by: Anton Potapov <anton.potapov@intel.com>

### CODE REVIEW

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst [hunk 1]
```diff
@@ -325,3 +325,66 @@ task:
         }
         tg.wait();
     }
+
+Deferred task creation
+----------------------
+The low-level task API of TBB separates task creation from actual spawning. This separation allows to
```
Comment (alexandraepan) [inline doc/main/tbb_userguide/Migration_Guide/Task_API.rst:331]:
```suggestion
The TBB low-level task API separates the task creation from the actual spawning. This separation allows to
```

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst [hunk 2]
```diff
@@ -325,3 +325,66 @@ task:
...
+postpone task spawning while having parent task blocked(and final result production) from premature leave.
```
Comment (alexandraepan) [inline doc/main/tbb_userguide/Migration_Guide/Task_API.rst:332]:
```suggestion
postpone the task spawning, while the parent task and final result production are blocked from premature leave. 
```

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst [hunk 3]
```diff
@@ -325,3 +325,66 @@ task:
...
+
+        tg.run(ChildTask{/*params*/});
+
+ 
+        tg.wait();
+        // Control flow will get here once both ChildTask CallBackTask are executed
+        // i.e. aftet the callback is called  
+    }
+
+Here ``oneapi::tbb::task_group::defer`` adds a new task into the ``tg``. Howether the task is not spawned until
```
Comment (alexandraepan) [inline doc/main/tbb_userguide/Migration_Guide/Task_API.rst:387]:
```suggestion
Here ``oneapi::tbb::task_group::defer`` adds a new task into the ``tg``. However, the task is not spawned until 
```

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst [hunk 4]
```diff
@@ -325,3 +325,66 @@ task:
...
+
+ 
+        tg.wait();
+        // Control flow will get here once both ChildTask CallBackTask are executed
+        // i.e. aftet the callback is called  
+    }
+
+Here ``oneapi::tbb::task_group::defer`` adds a new task into the ``tg``. Howether the task is not spawned until 
+``oneapi::tbb::task_arena::enqueue`` is called. 
+Please note that call to ``oneapi::tbb::task_group::wait`` will not return control untiol both ``ChildTask`` and
```
Comment (alexandraepan) [inline doc/main/tbb_userguide/Migration_Guide/Task_API.rst:389]: .. note::
  The call to ``oneapi::tbb::task_group::wait`` does not return the control until both ``ChildTask`` and ``CallBackTask`` 
  are executed.

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst [hunk 5]
```diff
@@ -325,3 +325,66 @@ task:
         }
         tg.wait();
     }
+
+Deferred task creation
+----------------------
+The low-level task API of TBB separates task creation from actual spawning. This separation allows to
+postpone task spawning while having parent task blocked(and final result production) from premature leave. 
+For example, ``RootTask``, ``ChildTask``, and ``CallBackTask`` are the user-side functors that
+inherit ``tbb::task`` and implement its interface. Then blocking of ``RootTask`` from premature leave
```
Comment (alexandraepan) [inline doc/main/tbb_userguide/Migration_Guide/Task_API.rst:334]:
```suggestion
inherit ``tbb::task`` and implement its interface. Then, blocking the ``RootTask`` from leaving prematurely
```

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst [hunk 6]
```diff
@@ -325,3 +325,66 @@ task:
...
         tg.wait();
     }
+
+Deferred task creation
+----------------------
+The low-level task API of TBB separates task creation from actual spawning. This separation allows to
+postpone task spawning while having parent task blocked(and final result production) from premature leave. 
+For example, ``RootTask``, ``ChildTask``, and ``CallBackTask`` are the user-side functors that
+inherit ``tbb::task`` and implement its interface. Then blocking of ``RootTask`` from premature leave 
+and waiting on it is implemented as:
```
Comment (alexandraepan) [inline doc/main/tbb_userguide/Migration_Guide/Task_API.rst:335]:
```suggestion
and waiting on it is implemented as follows: 
```

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst [hunk 7]
```diff
@@ -325,3 +325,66 @@ task:
...
+
+.. code:: cpp
+
+    #define TBB_PREVIEW_TASK_GROUP_EXTENSIONS 1
+    #include <oneapi/tbb/task_group.h>
+
+    int main(){
+        oneapi::tbb::task_group tg;
+        oneapi::tbb::task_arena arena;
+        // Assuming ChildTask, and CallBackTask are defined.
```
Comment (alexandraepan) [inline doc/main/tbb_userguide/Migration_Guide/Task_API.rst:371]:
```suggestion
        // Assuming ChildTask and CallBackTask are defined.
```

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst [hunk 8]
```diff
@@ -325,3 +325,66 @@ task:
...
+        root.set_ref_count(3);
+        
+        tbb::task::spawn(child);
+        
+        register_callback([cb_task&](){
+            tbb::task::enqueue(cb_task);
+        });
+
+        root.wait_for_all();
+        // Control flow will rich here only after both ChildTask and CallBackTask are executed,
```
Comment (alexandraepan) [inline doc/main/tbb_userguide/Migration_Guide/Task_API.rst:357]: not clear what it reaches, maybe you meant **Control flow is reached**

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst [hunk 9]
```diff
@@ -325,3 +325,66 @@ task:
...
+        // i.e. aftet the callback is called
```
Comment (alexandraepan) [inline doc/main/tbb_userguide/Migration_Guide/Task_API.rst:358]:
```suggestion
        // i.e. after the callback is called  
```

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst [hunk 10]
```diff
@@ -325,3 +325,66 @@ task:
...
+        
+        register_callback([&tg, c = std::move(cb), &arena]{
+            arena.enqueue(c);
+        });        
+
+        tg.run(ChildTask{/*params*/});
+
+ 
+        tg.wait();
+        // Control flow will get here once both ChildTask CallBackTask are executed
```
Comment (alexandraepan) [inline doc/main/tbb_userguide/Migration_Guide/Task_API.rst:383]:
```suggestion
        // Control flow gets here once both ChildTask and CallBackTask are executed
```

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst [hunk 11]
```diff
@@ -325,3 +325,66 @@ task:
...
+        // i.e. aftet the callback is called
```
Comment (alexandraepan) [inline doc/main/tbb_userguide/Migration_Guide/Task_API.rst:384]:
```suggestion
        // i.e. after the callback is called  
```

**Comment by alexandraepan** [comment]
hi @anton-potapov, please apply all of the comments (or let me know of you disagree with something), so I could approve it.

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst (hunk 4 above)
Comment (anton-potapov) [reply doc/main/tbb_userguide/Migration_Guide/Task_API.rst]: added   ``  .. note :: ``

**Comment by alexandraepan** [comment]
lgtm, still do not understand the sentence from the unresolved conversation, but the rest is just fine :)

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst (hunk 8 above)
Comment (anton-potapov) [reply doc/main/tbb_userguide/Migration_Guide/Task_API.rst:357]: This piece of programming jargon ([Control Flow](https://en.wikipedia.org/wiki/Control_flow) ) means that CPU executing thid program will reach here only when both ChildTask and CallBackTask are executed

**Review by kboyarinov — DISMISSED** [review]

File: doc/main/tbb_userguide/Migration_Guide/Task_API.rst (hunk 8 above)
Comment (alexandraepan) [reply doc/main/tbb_userguide/Migration_Guide/Task_API.rst:420]: Ok, I see now, thanks, then change it to reach instead of rich

Comment (anton-potapov) [reply doc/main/tbb_userguide/Migration_Guide/Task_API.rst]: channged

**Review by omalyshe — APPROVED** [review]

**Pull request merged**